PlayerCollection.cs

```diff
@@ -60,9 +60,12 @@ public void AddPlayer(CommonFields.Player arcPlayer, Func<uint, uint, AsyncTextu
 
         public void UpdatePlayerSpecialization(string characterName, uint newSpec)
         {
-            var hasPlayer = _players.TryGetValue(characterName, out var player);
-            if (!hasPlayer) return;
-            if (player.CurrentSpecialization == newSpec) return;
+            var allPlayers = _players.Values.Concat(
+                _players.ToList().SelectMany(player => player.Value.PreviouslyPlayedCharacters)
+            );
+            var player = allPlayers.FirstOrDefault(player => player.CharacterName == characterName);
+            if (player == null) return;
+            if (player?.CurrentSpecialization == newSpec) return;
```

Comment:
You don't need the `?` anymore since you null-check the line above!
```suggestion
            if (player.CurrentSpecialization == newSpec) return;
```